### PR TITLE
Add the `io.linkerd.proxy.inbound` API

### PR DESCRIPTION
**Pull request by olix0r**
The `io.linkerd.proxy.inbound` package provides the `InboundServerDiscovery`
service to support discovering inbound proxy configurations to support
access policies.

### CODE REVIEW

**Comment by olix0r** [comment]
Help me bikeshed naming. Does this inbound / InboundServerDiscovery make sense? Do the names make sense generally?

**Review by adleong — APPROVED** [review]
The API seems correct and intuitive from reading it.

I don't have any strong feelings about naming and the only questionable one is `InboundServerDiscovery` just because it sounds so similar to "Service Discovery".  The word "Discovery" feels a bit weird to use on the inbound side of things, even though it does parallel the outbound service discovery.  What about something like "InboundConfiguration" or "InboundPolicy"?

With all that said, I think the name of the gRPC service is probably less important than e.g. the names of the methods or structs so I'm not too worried about it.

**Review by alpeb — COMMENTED** [review]
API looks good to me :+1: 
Given the controller is in Rust, the go code generated here is just for completeness sake, in case an external party wants to leverage this?

**Comment by olix0r** [comment]
> Given the controller is in Rust, the go code generated here is just for completeness sake

Yeah, I view this repo as the client/server bindings for these APIs in both Rust and Go. Ideally we wouldn't need to manage the generated Go files for any of this but that's a larger change. (It may be desirable to use the Go bindings for CLI tooling like the `endpoints` helper...)

**Review by alpeb — APPROVED** [review]

File: proto/inbound.proto [hunk 1]
```diff
@@ -0,0 +1,149 @@
+syntax = "proto3";
+
+package io.linkerd.proxy.inbound;
+
+option go_package = "github.com/linkerd/linkerd2-proxy-api/go/inbound";
+
+import "google/protobuf/duration.proto";
+import "net.proto";
+
+service InboundServerDiscovery {
```
Comment (olix0r) [inline proto/inbound.proto:10]: @adleong wdyt?

```suggestion
service InboundServerPolicies {
```

**Pull request merged**